src/swarm-tui/Swarm/TUI/Controller/EventHandlers.hs

```diff
@@ -0,0 +1,131 @@
+--
+-- The high level overview is this:
+-- 1. The 'SwarmEvent' is a enumeration of abstracts key events ('PauseEvent', etc.)
+-- 2. The 'AppState' definition contains the key configuration and dispatchers (keys to handlers)
+-- 3. Here in 'Swarm.TUI.Controller.EventHandlers' we declare the handlers for abstract events
+--    and also some non-customizable key handlers (e.g. escape and enter).
+-- 4. When provided with 'KeyConfig' (can include customized keybindings) we can
+--    'createKeyDispatchers' in 'Swarm.TUI.Model.StateUpdate' and store them in 'AppState'.
+-- 5. Finally in 'Swarm.TUI.Controller' the Brick event handler calls the stored dispatchers.
+
+-- | Create key dispatchers that call (abstract) event handlers based on given key config.
+--
+-- Fails if any key events have conflict within one dispatcher or when a main dispatcher
+-- has conflict with one of the subdispatchers.
+createKeyDispatchers ::
+  (Has (Throw SystemFailure) sig m) =>
+  KeyConfig SwarmEvent ->
+  m SwarmKeyDispatchers
+createKeyDispatchers config = do
+  mainGameDispatcher <- buildDispatcher mainEventHandlers
+  let buildSubMainDispatcher = buildSubDispatcher "Main game events" mainGameDispatcher
+  replDispatcher <- buildSubMainDispatcher "REPL panel events" replEventHandlers
+  worldDispatcher <- buildSubMainDispatcher "World view panel events" worldEventHandlers
+  robotDispatcher <- buildSubMainDispatcher "Robot inventory panel events" robotEventHandlers
+  return SwarmKeyDispatchers {..}
+ where
+  -- this error handling code is modified version of the brick demo app:
+  -- https://github.com/jtdaugherty/brick/blob/764e66897/programs/CustomKeybindingDemo.hs#L216
+  buildDispatcher handlers = case keyDispatcher config handlers of
+    Left collisions ->
+      throwLoadingFailure $
+        "Error: some key events have the same keys bound to them.\n"
+          : handlerErrors collisions
+    Right d -> return d
+  buildSubDispatcher parentName parentDispatcher name handlers = do
+    d <- buildDispatcher handlers
+    let collisions = conflicts parentDispatcher d
+    if null collisions
+      then return d
+      else
+        throwLoadingFailure $
+          ("Error: some key events have keys bound to them in '" <> parentName <> "' and in '" <> name <> "'")
+            : handlerErrors collisions
+
+  throwLoadingFailure = throwError . CustomFailure . T.intercalate "\n"
+  handlerErrors collisions = flip map collisions $ \(b, hs) ->
+    let hsm = "Handlers with the '" <> BK.ppBinding b <> "' binding:"
+        hss = flip map hs $ \h ->
+          let trigger = case BK.kehEventTrigger $ BK.khHandler h of
+                ByKey k -> "triggered by the key '" <> BK.ppBinding k <> "'"
+                ByEvent e -> "triggered by the event '" <> fromMaybe "<unknown>" (BK.keyEventName swarmEvents e) <> "'"
+              desc = BK.handlerDescription $ BK.kehHandler $ BK.khHandler h
+           in "  " <> desc <> " (" <> trigger <> ")"
```

Comment:
Good opportunity to use the [`squote`](https://github.com/swarm-game/swarm/blob/9b1b6985fa51ca396358cb77c7d496f807eafae8/src/swarm-util/Swarm/Util.hs#L430) and [`parens`](https://github.com/swarm-game/swarm/blob/9b1b6985fa51ca396358cb77c7d496f807eafae8/src/swarm-util/Swarm/Util.hs#L442) functions

Reply:
I wish we had f-strings `f"event '{event}'"`, they would make this much easier to read.

At least I can drop `fromMaybe`.

Reply:
https://hackage.haskell.org/package/fmt ?